File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        // approach: two passes, first to make hashmap of old node : copied node,
        // second to connect ptrs
        
        Map<Node, Node> hashmap = new HashMap<>();
        hashmap.put(null, null); // edge case
        
        Node curr = head;
        
        while (curr != null) {
            Node copy = new Node(curr.val);
            hashmap.put(curr, copy);
            curr = curr.next;
        }
        
        curr = head;
        
        while (curr != null) {
            Node copy = hashmap.get(curr);
            copy.next = hashmap.get(curr.next);
            copy.random = hashmap.get(curr.random);
            curr = curr.next;
        }
        
        return hashmap.get(head);
    }
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        # approach: hashmap of old node : new node, two passes
        # 1st pass: add nodes to hashmap
        # 2nd pass: connect ptrs
        
        hashmap = { None : None }
        
        curr = head
        while curr:
            hashmap[curr] = Node(curr.val)
            curr = curr.next
        
        curr = head
        while curr:
            hashmap[curr].next = hashmap[curr.next]
            hashmap[curr].random = hashmap[curr.random]
            curr = curr.next
            
        return hashmap[head]